### PR TITLE
[device_info] Port device_info to use platform interface

### PR DESCRIPTION
**Pull request by yash1200**
## Description

In this PR, I have ported `device_info` to use `device_info_platform_package` .

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. [shared_preferences]
- [x] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://www.dartlang.org/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

**Comment by yash1200** [comment]
Hey @ditman , kindly have a look at this one. :)

File: packages/device_info/device_info/lib/device_info.dart [hunk 1]
```diff
@@ -33,292 +30,5 @@ class DeviceInfoPlugin {
...
   /// See: https://developer.apple.com/documentation/uikit/uidevice
   Future<IosDeviceInfo> get iosInfo async =>
-      _cachedIosDeviceInfo ??= IosDeviceInfo._fromMap(
-          await channel.invokeMapMethod<String, dynamic>('getIosDeviceInfo'));
-}
-
-/// Information derived from `android.os.Build`.
-///
-/// See: https://developer.android.com/reference/android/os/Build.html
-class AndroidDeviceInfo {
```
Comment (ditman) [inline packages/device_info/device_info/lib/device_info.dart:43]: All these classes that you moved from device_info to device_info_platform_interface need to be re-exported (so they're still part of the public API). 

Look at the current documentation of device_info to know which classes you need to ensure are visible from the plugin: https://pub.dev/documentation/device_info/latest/device_info/device_info-library.html

File: packages/device_info/device_info/example/lib/main.dart [hunk 2]
```diff
@@ -11,6 +11,9 @@ import 'package:flutter/material.dart';
 import 'package:flutter/services.dart';
 import 'package:device_info/device_info.dart';
 
+import 'package:device_info_platform_interface/device_info_platform_interface.dart';
+import 'package:device_info_platform_interface/model/android_device_info.dart';
```
Comment (ditman) [inline packages/device_info/device_info/example/lib/main.dart:15]: Don't import from the model. This should come from the device_info_platform_interface.dart file.

Comment (ditman) [reply packages/device_info/device_info/example/lib/main.dart:15]: Here: https://github.com/flutter/plugins/blob/master/packages/device_info/device_info_platform_interface/lib/device_info_platform_interface.dart#L14

File: packages/device_info/device_info/example/test_driver/device_info_e2e.dart [hunk 3]
```diff
@@ -7,6 +7,9 @@ import 'package:flutter_test/flutter_test.dart';
 import 'package:device_info/device_info.dart';
 import 'package:e2e/e2e.dart';
 
+import 'package:device_info_platform_interface/device_info_platform_interface.dart';
+import 'package:device_info_platform_interface/model/android_device_info.dart';
```
Comment (ditman) [inline packages/device_info/device_info/example/test_driver/device_info_e2e.dart:11]: Same comment, this import is unnecessary.

File: packages/device_info/device_info/lib/device_info.dart (hunk 1 above)
Comment (yash1200) [reply packages/device_info/device_info/lib/device_info.dart:43]: Completely forgot about that. :( . I've done the changes now.

File: packages/device_info/device_info/example/test_driver/device_info_e2e.dart [hunk 4]
```diff
@@ -7,9 +7,6 @@ import 'package:flutter_test/flutter_test.dart';
 import 'package:device_info/device_info.dart';
 import 'package:e2e/e2e.dart';
 
-import 'package:device_info_platform_interface/device_info_platform_interface.dart';
```
Comment (ditman) [inline packages/device_info/device_info/example/test_driver/device_info_e2e.dart]: My bad, you still need the `device_info_platform_interface.dart` import:

`import 'package:device_info_platform_interface/device_info_platform_interface.dart';`

(But that's the *only* import you should need from the device_info_platform_interface package here.)

File: packages/device_info/device_info/lib/device_info.dart [hunk 5]
```diff
@@ -3,28 +3,26 @@
 // found in the LICENSE file.
 
 import 'dart:async';
+import 'package:device_info_platform_interface/device_info_platform_interface.dart';
 
-import 'package:flutter/services.dart';
+export 'package:device_info_platform_interface/model/android_device_info.dart';
+export 'package:device_info_platform_interface/model/ios_device_info.dart';
```
Comment (ditman) [inline packages/device_info/device_info/lib/device_info.dart:9]: Ah, forgot about the current syntax for this, so you don't have to access "private" parts of the platform_interface package. Do it like this:

`export 'package:device_info_platform_interface/device_info_platform_interface.dart' show ClassName1, ClassName2, ClassName3;`

Apologies, I should have been clearer. You can see this in other plugins, like [this extreme case in google_maps_flutter](https://github.com/flutter/plugins/blob/82f8f490cfa4b257fc987cd40bab983a8c1ab2e6/packages/google_maps_flutter/google_maps_flutter/lib/google_maps_flutter.dart#L20-L46).

Please, make sure you're only re-exporting the classes that you moved to the platform_interface.

The problem with the current version is that if we add by accident some code to those files, it gets reexported automatically, whether we want it or not.

Comment (yash1200) [reply packages/device_info/device_info/lib/device_info.dart]: No need for an apology sir, I am learning many things from your reviews and it's a great learning this way. BTW I have pushed the changes.

**Comment by yash1200** [comment]
Thank You Sir!

**Comment by yash1200** [comment]
Hey @ditman , if I were to add support for linux in `device_info` , then what are the necessary stats that I must include in it. Should I include all the stats which `neofetch` shows?

**Comment by ditman** [comment]
@yash1200 I'm not a desktop expert. Maybe @stuartmorgan has a better idea. I'd say: start with the stats that you need for your app, and build up from there.

**Comment by stuartmorgan** [comment]
I'm not sure there's really any way for me to answer that question; even with just two implementations (iOS and Android) there's almost nothing at all in common between the two, so it's hard to judge what is or is not in scope. Perhaps before extending this to N new platforms there should be a clearer articulation by the ecosystem team of what general criteria should be used to decide what is added to it.

(This PR is probably not the place for that discussion though.)

**Comment by stuartmorgan** [comment]
Filed https://github.com/flutter/flutter/issues/64569

**Pull request merged**